### DISCUSSION TITLE
Route could be more semantic

### DISCUSSION THREAD

**Discussion by lpj145** [loco-rs/loco#363] [Ideas]
## Feature Request
Sorry to boter you guys (again!)
**Background:**
I have worked with some top http frameworks like: cakephp, django, express, nest, fastify and elysia(bun) and I had to say that: any of those have incredible things to aggregate value to this project, should we only stay close to rails?

**Describe the solution you'd like**
PS* I've written this code on github interface I don't know if it will compile
```rust
struct HttpOptions {
/// can we call this feature -> feature flag typo
    permissions: Vec<String>,
    is_public: bool,
}

trait HttpHooks {
    onResponse: (req: Request, resp: Response) -> Response
}

struct Route {}

impl Route {
    pub fn get<T>(mut self, path: String, handle: () -> Result<IntoResponse<T>>, options: dyn HttpOptions) -> Self {}
}

/// Then uses the Route like that
pub fn routes() -> Routes {
    Routes::new().prefix("user").get("/current", defined_top_function, HttpOptions { is_public: true, permissions: vec!("ADMIN")})
}
```

I know about the part of dynamic dispatch but we can improve this, I see that this will be more verbose than the old fashion way, but I don't see if we will be good to keep the auth: Auth as a argument of my handle http request function.

By the way we are fighting against trade-offs what is best choice

1. Put those arguments to translate some more http behavior things in that way we should write a FromParts extractor if I want to put permission on that function.
2. Describe the route as I did passing all the meta information to the Router
3. Or just use proc_macros like #[authorize, permissions] todo this kind of thing.

I can't see right now what is the best way to go.

**Describe alternatives you've considered**

Create my own Router and uses that on top of axum.

**Comment by lpj145** [comment]
With this approach what can I achieve:
- Describe api's with meta information wiches could be usable to reproduce and generate apis ( [see how fastify describe routes](https://fastify.dev/docs/latest/Reference/Routes/#routes-options))
- Easily found any endpoint handle using a simple ctrl+shift+f on vscode on all files]
- Steps up to reproduce a good openapi schema

Right now I just want to lock all my endpoints with auth, as I can see I could use Auth as argument of every function, but this could be easily achieve by adding a auth middleware, and in my tests I can use my handle http function

**Comment by jondot** [comment]
Thanks! we always want to see better ways to developer experience!
I think if you can come up with a branch that showcases an incremental (prefer non breaking) improvement to our routes it would be fantastic!
I'll also convert this issue to discussions so we can continue there ❤️
